tools/aux/tpm2_print.c

```diff
@@ -0,0 +1,279 @@
+        goto error;
+    }
+
+    // print digest in hex
+    printf("%s.pcrDigest=", prefix);
+    if(!print_hex(fd, filename, digest_size)) {
+        goto error;
+    }
+
+    // success
+    return true;
+
+    read_error:
+    FILE_2_SHORT_ERROR(filename);
+    error:
+    return false;
+}
+
+static bool print_TPMS_ATTEST(FILE* fd, const char* const filename) {
+    // print magic without converting endianness
+    UINT32 magic;
+    if(!files_read_bytes(fd, (UINT8*)&magic, sizeof(UINT32))) {
+        goto read_error;
+    }
+    printf("TPMS_ATTEST.magic=");
+    print_hex_buffer((const UINT8*)&magic, sizeof(UINT32));
+    magic = tpm2_util_ntoh_32(magic); // finally, convert endianness
+
+    // check magic
+    if(magic != TPM_GENERATED_VALUE) {
+        LOG_ERR("%s: Bad magic", filename);
+        goto error;
+    }
+
+    UINT16 type;
+    if(!files_read_bytes(fd, (UINT8*)&type, sizeof(UINT16))) {
+        goto read_error;
+    }
+    printf("TPMS_ATTEST.type=");
+    print_hex_buffer((const UINT8*)&type, sizeof(UINT16));
+    type = tpm2_util_ntoh_16(type); // finally, convert endianness
+
+    printf("TPMS_ATTEST.qualifiedSigner=");
+    if(!print_tpm2b_hex(fd, filename)) {
+        goto error;
+    }
+
+    printf("TPMS_ATTEST.extraData=");
+    if(!print_tpm2b_hex(fd, filename)) {
+        goto error;
+    }
+
+    if(!print_clock_info(fd, filename, "TPMS_ATTEST")) {
+        goto error;
+    }
+
+    printf("TPMS_ATTEST.firmwareVersion=");
+    print_hex(fd, filename, sizeof(UINT64));
+
+    switch(type) {
+    case TPM_ST_ATTEST_QUOTE:
+        if(!print_TPMS_QUOTE_INFO(fd, filename, "TPMS_ATTEST.attested.quote")) {
+            goto error;
+        }
+        break;
+
+    default:
+        LOG_ERR("%s: Cannot print unsupported type 0x%x", filename, (unsigned int)type);
+        goto error;
+    }
+
+    // success
+    return true;
+
+    read_error:
+    FILE_2_SHORT_ERROR(filename);
+    error:
+    return false;
+}
+
+int main(int argc, char *argv[]) {
+    const char* object_type = NULL;
+    const char* filename = NULL;
+    FILE* fd = NULL;
+
+    if(argc == 3) {
+        object_type = argv[1];
+        filename = argv[2];
+    }
+    else {
+        LOG_ERR("Usage: tpm2_print <object_type> <filename>");
+        goto error;
+    }
+
+    printf("filename=%s\n", filename);
+    fd = fopen(filename, "rb");
+    if(!fd) {
+        LOG_ERR("%s: Could not open file", filename);
+        goto error;
+    }
+
+    if(strcmp(object_type, "TPMS_ATTEST") == 0) {
+        if(!print_TPMS_ATTEST(fd, filename)) {
+            goto error;
+        }
+    }
+    else {
+        LOG_ERR("Unsupported object_type %s; only TPMS_ATTEST is presently supported", object_type);
+        goto error;
+    }
+
+    // print number of bytes read
+    printf("ftell=%ld\n", ftell(fd));
+
+    // check EOF by reading one more byte
+    UINT8* garbage;
+    int fread_result = fread(&garbage, 1, 1, fd);
```

Comment:
Could you re-work this logic with files_load_bytes_from_file_or_stdin? It would require you to know how much data you need upfront.

Reply:
Dropped per comments below.